src/theory/quantifiers/sygus_inference.cpp

```diff
@@ -0,0 +1,188 @@
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2017 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of sygus_inference
+ **/
+
+#include "theory/quantifiers/sygus_inference.h"
+#include "theory/quantifiers/quantifiers_attributes.h"
+
+using namespace std;
+using namespace CVC4::kind;
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+SygusInference::SygusInference() {}
+
+bool SygusInference::simplify(std::vector<Node>& assertions)
+{
+  Trace("sygus-infer") << "Sygus inference : " << std::endl;
+
+  if (assertions.empty())
+  {
+    Trace("sygus-infer") << "...fail: empty assertions." << std::endl;
+    return false;
+  }
+
+  NodeManager* nm = NodeManager::currentNM();
+
+  // collect free variables in all assertions
+  std::vector<Node> qvars;
+  std::map<TypeNode, std::vector<Node> > qtvars;
+  std::vector<Node> free_functions;
+
+  std::vector<TNode> visit;
+  std::unordered_set<TNode, TNodeHashFunction> visited;
+
+  std::vector<Node> processed_assertions;
+  for (const Node& as : assertions)
+  {
+    Trace("sygus-infer") << "  " << as << std::endl;
+    // substitution for this assertion
+    std::vector<Node> vars;
+    std::vector<Node> subs;
+    std::map<TypeNode, unsigned> type_count;
+    Node pas = as;
+    if (as.getKind() == FORALL)
+    {
+      pas = as[1];
+      // infer prefix
+      for (const Node& v : as[0])
+      {
+        TypeNode tnv = v.getType();
+        unsigned vnum = type_count[tnv];
+        type_count[tnv]++;
+        if (vnum < qtvars[tnv].size())
+        {
+          vars.push_back(v);
+          subs.push_back(qtvars[tnv][vnum]);
+        }
+        else
+        {
+          Assert(vnum == qtvars[tnv].size());
+          qtvars[tnv].push_back(v);
+          qvars.push_back(v);
+        }
+      }
+      if (!vars.empty())
+      {
+        pas =
+            pas.substitute(vars.begin(), vars.end(), subs.begin(), subs.end());
+      }
+    }
+
+    // collect free functions, ensure no quantified formulas
+    TNode cur = pas;
+    // compute free variables
+    visit.push_back(cur);
+    do
+    {
+      cur = visit.back();
+      visit.pop_back();
+      if (visited.find(cur) == visited.end())
+      {
+        visited.insert(cur);
+        if (cur.getKind() == APPLY_UF)
+        {
+          Node op = cur.getOperator();
+          if (std::find(free_functions.begin(), free_functions.end(), op)
+              == free_functions.end())
+          {
+            free_functions.push_back(op);
+          }
+        }
+        else if (cur.getKind() == FORALL)
+        {
+          Trace("sygus-infer")
+              << "...fail: non-top-level quantifier." << std::endl;
```

Comment:
Add an issue to be robust to such cases. This function could prenex the formula as well as assume that skolemization has been performed (another issue to allow this inference to happen after preprocessing).

Reply:
We are now robust to preskolemization and prenexing after my latest commit.